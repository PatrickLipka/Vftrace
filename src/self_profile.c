#include <stdio.h>

#ifdef _SELF_PROFILE_VFTRACE
char *vftr_self_prof_filename = NULL;
FILE *vftr_self_prof_iohandle = NULL;
char *vftr_self_prof_iobuffer = NULL;
#endif
