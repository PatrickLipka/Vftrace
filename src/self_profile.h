#ifndef SELF_PROFILE_H
#define SELF_PROFILE_H

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <time.h>

#ifdef _SELF_PROFILE_VFTRACE
extern char *vftr_self_prof_filename;
extern FILE *vftr_self_prof_iohandle;
extern char *vftr_self_prof_iobuffer;

#define GET_SELF_PROF_FILENAME \
   do { \
      int pid = getpid(); \
      char *filename_base = "vftrace_self_profile-"; \
      int total_str_len = snprintf(NULL, 0, "%s%d", filename_base, pid) + 1; \
      vftr_self_prof_filename = (char*) malloc(total_str_len*sizeof(char)); \
      snprintf(vftr_self_prof_filename, \
               total_str_len, "%s%d", \
               filename_base, pid); \
   } while(0)

#define GET_SELF_PROF_FILEHANDLE \
   do { \
      vftr_self_prof_iohandle = \
         fopen(vftr_self_prof_filename, "w"); \
      size_t bufsize = 256*1024*1024; \
      vftr_self_prof_iobuffer = (char*) malloc(bufsize); \
      memset((void*) vftr_self_prof_iobuffer, 0, bufsize); \
      int status = setvbuf(vftr_self_prof_iohandle, \
                           vftr_self_prof_iobuffer, \
                           _IOFBF, bufsize); \
   } while(0)

#define INIT_SELF_PROF_VFTRACE \
   do { \
      GET_SELF_PROF_FILENAME; \
      GET_SELF_PROF_FILEHANDLE; \
   } while(0)

#define FREE_SELF_PROF_FILENAME \
   do { \
      if (vftr_self_prof_filename != NULL) { \
         free(vftr_self_prof_filename); \
         vftr_self_prof_filename = NULL; \
      } \
   } while(0)

#define CLOSE_SELF_PROF_FILEHANDLE \
   do { \
      if (vftr_self_prof_iohandle != NULL) { \
         fclose(vftr_self_prof_iohandle); \
         vftr_self_prof_iohandle = NULL; \
         free(vftr_self_prof_iobuffer); \
         vftr_self_prof_iobuffer = NULL; \
      } \
   } while(0)
   
#define FINALIZE_SELF_PROF_VFTRACE \
   do { \
      CLOSE_SELF_PROF_FILEHANDLE; \
      FREE_SELF_PROF_FILENAME; \
   } while(0)

#define GET_TIMESTAMP \
   struct timespec timestamp; \
   clock_gettime(CLOCK_MONOTONIC, &timestamp);

#define SELF_PROFILE_START_FUNCTION \
   do { \
      GET_TIMESTAMP; \
      fprintf(vftr_self_prof_iohandle, \
              "Enter: %s at %ld s %ld ns\n", \
              __FUNCTION__, \
              timestamp.tv_sec, \
              timestamp.tv_nsec); \
   } while(0)
   

#define SELF_PROFILE_END_FUNCTION \
   do { \
      GET_TIMESTAMP; \
      fprintf(vftr_self_prof_iohandle, \
              "Leave: %s at %ld s %ld ns\n", \
              __FUNCTION__, \
              timestamp.tv_sec, \
              timestamp.tv_nsec); \
   } while(0)

#else

#define INIT_SELF_PROF_VFTRACE
#define SELF_PROFILE_START_FUNCTION
#define SELF_PROFILE_END_FUNCTION
#define FINALIZE_SELF_PROF_VFTRACE

#endif // ifdef SELF_PROFILE_VFTRACE

#endif // define SELF_PROFILE_H
