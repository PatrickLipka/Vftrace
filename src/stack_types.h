#ifndef STACK_TYPES_H
#define STACK_TYPES_H

#include <stdint.h>
#include <stdbool.h>

#include "custom_types.h"
#include "profiling_types.h"

typedef struct {
   // address of the function
   uintptr_t address;
   // is this function measured precisely?
   bool precise;
   // pointer to calling stack
   int caller;
   // pointers to called functions
   int maxcallees;
   int ncallees;
   int *callees;
   // local stack ID
   int lid;
   // global stack ID
   int gid;
   // name of function on top of stack
   char *name;
   // cleaned (module striped, or demangled) name
   char *cleanname;
   // profiling data
   profilelist_t profiling;
   // Data that is filled in during finalization
   uint64_t hash;
} stack_t;

typedef struct {
   int nstacks;
   int maxstacks;
   stack_t *stacks;
} stacktree_t;

#endif
