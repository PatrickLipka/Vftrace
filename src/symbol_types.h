#ifndef SYMBOL_TYPES_H
#define SYMBOL_TYPES_H

#include <stdint.h>
#include <sys/types.h>

#include "custom_types.h"

typedef struct {
   uintptr_t addr;
   char *name; /* Not de-mangled function name */
   int index; /* Section index */
} symbol_t;

typedef struct {
   unsigned int nsymbols;
   symbol_t *symbols;
} symboltable_t;

// types to contain library paths
typedef struct {
   off_t base;
   off_t offset;
   char *path;
} library_t;

typedef struct {
   int nlibraries;
   int maxlibraries;
   library_t *libraries;
} librarylist_t;

#endif
