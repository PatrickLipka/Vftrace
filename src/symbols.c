#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>

#include <string.h>
#include <elf.h>

#include "address_type.h"
#include "symbols.h"
#include "ElfFormat.h"
#include "sorting.h"

library_t vftr_parse_maps_line(char *line) {
   library_t library = {
      .base = 0,
      .offset = 0,
      .path = NULL
   };
   // a line has the following format:
   // <baseaddr>-<topaddr> <permissions> <offset> <device> <inode> <libpath>
   // split the string by tokens
   char *base = strtok(line,"-");
   // skip over topaddr
   strtok(NULL," ");
   char *permissions = strtok(NULL," ");
   // only continue parsing if the library is marked executable
   if (permissions[2] != 'x') {return library;}
   char *offset = strtok(NULL, " ");
#ifndef __VMAP_OFFSET
   if (strncmp(offset, "00000000", 8)) {return library;}
#endif
   // skip over devide
   strtok(NULL," ");
   // skip over inode
   strtok(NULL," ");
   char *path = strtok(NULL," \n");
   // only continue if a valid path is found
   if (path == NULL || path[0] == '[') {return library;}
   // Filter out devices and system libraries
   // They are not instrumented and can be discarded
   // test for common path names
   if (!strncmp(path, "/dev/", 5)) {return library;}
   if (!strncmp(path, "/usr/lib", 8)) {return library;}
   if (!strncmp(path, "/usr/lib64", 10)) {return library;}
   if (!strncmp(path, "/lib", 4)) {return library;}
#ifdef __ve__
   if (!strncmp(path, "/opt/nec/ve/veos/lib", 20)) {return library;}
#endif
   sscanf(base, "%ld", &(library.base));
   library.path = strdup(path);
#ifdef __VMAP_OFFSET
   library.offset = strtoul(offset, NULL, 16);
#else
   library.offset = 0L;
#endif
   return library;
}

void vftr_print_library_list(FILE *fp, librarylist_t librarylist) {
   fprintf(fp, "Found libraries:\n");
   for (int ilib=0; ilib<librarylist.nlibraries; ilib++) {
      fprintf(fp, "%s (base=%lu, offset=%lu)\n",
              librarylist.libraries[ilib].path,
              librarylist.libraries[ilib].base,
              librarylist.libraries[ilib].offset);
   }
   fprintf(fp, "\n");
}

// read the different library paths from the applications map
librarylist_t vftr_read_library_maps() {
   char *fmappath = "/proc/self/maps";
   FILE *fmap = fopen(fmappath, "r");
   if (fmap == NULL) {
      perror (fmappath);
      abort();
   }

   librarylist_t librarylist = {
      .nlibraries = 0,
      .maxlibraries = 0,
      .libraries = NULL
   };

   char *lineptr = NULL;
   size_t buffsize = 0;
   ssize_t readbytes = 0;
   // read all lines
   while ((readbytes = getline(&lineptr, &buffsize, fmap)) > 0) {
      library_t library = vftr_parse_maps_line(lineptr);
      if (library.path != NULL) {
         // append it to the librarylist
         int idx = librarylist.nlibraries;
         librarylist.nlibraries++;
         if (librarylist.nlibraries > librarylist.maxlibraries) {
            librarylist.maxlibraries = 1.4*(librarylist.maxlibraries+1);
            librarylist.libraries = (library_t*)
               realloc(librarylist.libraries, librarylist.maxlibraries*sizeof(library_t));
         }
         librarylist.libraries[idx] = library;
      }
   }
   free(lineptr);
   fclose(fmap);
#ifdef _DEBUG
   vftr_print_library_list(stderr, librarylist);
#endif
   return librarylist;
}

void vftr_free_librarylist(librarylist_t *librarylist_ptr) {
   librarylist_t librarylist = *librarylist_ptr;
   if (librarylist.nlibraries > 0) {
      for (int ilib=0; ilib<librarylist.nlibraries; ilib++) {
         free(librarylist.libraries[ilib].path);
         librarylist.libraries[ilib].path = NULL;
      }
      free(librarylist.libraries);
      librarylist.libraries = NULL;
   }
}

void vftr_print_symbol_table(FILE *fp, symboltable_t symboltable) {
   for (unsigned int isym=0; isym<symboltable.nsymbols; isym++) {
      fprintf(fp, "%d 0x%llx %s\n",
              symboltable.symbols[isym].index,
              (unsigned long long int)
              symboltable.symbols[isym].addr,
              symboltable.symbols[isym].name);
   }
   fprintf(fp, "\n");
}

symboltable_t vftr_read_symbols() {
   // get all library paths that belong to the program
   librarylist_t librarylist = vftr_read_library_maps();

   symboltable_t symboltable = {
      .nsymbols = 0,
      .symbols = NULL,
   };
   // loop over all libraries and get their symbol tables
   for (int ilib=0; ilib<librarylist.nlibraries; ilib++) {
      FILE *fp = fopen(librarylist.libraries[ilib].path, "r");
      if (fp == NULL) {
         perror(librarylist.libraries[ilib].path);
         abort();
      }
      Elf64_Ehdr ElfHeader = vftr_read_elf_header(fp);
      Elf64_Shdr *ElfSectionHeader = vftr_read_elf_section_header(fp, ElfHeader);
      char *header_strtab = vftr_read_elf_header_string_table(fp, ElfHeader,
                                                              ElfSectionHeader);
      if (header_strtab == NULL) {
         free(ElfSectionHeader);
         fclose(fp);
         continue;
      }
      int strtabidx = vftr_get_elf_string_table_index(header_strtab,
                                                      ElfHeader,
                                                      ElfSectionHeader);
      if (strtabidx == -1) {
#ifdef _DEBUG
         fprintf(stderr, "No symbol string table in %s\n", librarylist.libraries[ilib].path);
#endif
         free(ElfSectionHeader);
         free(header_strtab);
         fclose(fp);
         continue;
      }
      char *stringtab = vftr_read_elf_symbol_string_table(fp, ElfSectionHeader,
                                                          strtabidx);
      int symtabidx = vftr_get_elf_symbol_table_index(header_strtab,
                                                      ElfHeader,
                                                      ElfSectionHeader);
      if (symtabidx == -1) {
#ifdef _DEBUG
         fprintf(stderr, "No symbol table in %s\n", librarylist.libraries[ilib].path);
#endif
         free(ElfSectionHeader);
         free(header_strtab);
         free(stringtab);
         fclose(fp);
         continue;
      }

      Elf64_Sym *ElfSymbolTable = vftr_read_elf_symbol_table(fp, ElfSectionHeader,
                                                             symtabidx);



      int symbolCount = 0;
      int validSymbolCount = 0;
      vftr_elf_symbol_table_count(ElfSectionHeader, symtabidx, ElfSymbolTable,
                                  &symbolCount, &validSymbolCount);
      if (validSymbolCount > 0) {
         // append the symbols from this library to the symboltable
         int nsymold = symboltable.nsymbols;
         int nsymnew = nsymold + validSymbolCount;
         symboltable.symbols = (symbol_t*)
            realloc(symboltable.symbols, nsymnew*sizeof(symbol_t));
         int copiedsymbols = 0;
         for (int isymb=0; isymb<symbolCount; isymb++) {
            Elf64_Sym s = ElfSymbolTable[isymb];
            if (ELF64_ST_TYPE(s.st_info) == STT_FUNC && s.st_value) {
               int jsymb = nsymold + copiedsymbols;
               copiedsymbols++;
#if defined(__ve__)
               symboltable.symbols[jsymb].addr =
                  (uintptr_t) (s.st_value);
#else
               off_t base = librarylist.libraries[ilib].base;
               symboltable.symbols[jsymb].addr =
                  (uintptr_t) (base + s.st_value);
#endif
               // Copy symbol name
               symboltable.symbols[jsymb].name = strdup(stringtab+s.st_name);
               symboltable.symbols[jsymb].index = s.st_shndx;
            }
         }
         symboltable.nsymbols = nsymnew;
      }

      free(ElfSectionHeader);
      free(header_strtab);
      free(stringtab);
      free(ElfSymbolTable);
      fclose(fp);
   }

   // sort the symbol table for faster access lateron with a binary search
   vftr_radixsort_symboltable(symboltable.nsymbols,
                              symboltable.symbols);

   vftr_free_librarylist(&librarylist);
#ifdef _DEBUG
   fprintf(stderr, "Elf-Symboltable:\n");
   vftr_print_symbol_table(stderr, symboltable);
#endif
   return symboltable;
}

void vftr_symboltable_free(symboltable_t *symboltable_ptr) {
   symboltable_t symboltable = *symboltable_ptr;
   if (symboltable.nsymbols > 0) {
      for (unsigned int isym=0; isym<symboltable.nsymbols; isym++) {
         free(symboltable.symbols[isym].name);
         symboltable.symbols[isym].name = NULL;
      }
      free(symboltable.symbols);
      symboltable.nsymbols = 0;
   }
}
