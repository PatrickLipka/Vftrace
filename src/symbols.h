#ifndef SYMBOLS_H
#define SYMBOLS_H

#include <sys/types.h>

typedef struct {
   unsigned long long addr;
   char *name; /* Not de-mangled function name */
   int index; /* Section index */
} symbol_t;

typedef struct {
   unsigned int nsymbols;
   symbol_t *symbols;
} symboltable_t;

// types to contain library paths
typedef struct {
   off_t base;
   off_t offset;
   char *path;
} library_t;

typedef struct {
   int nlibraries;
   int maxlibraries;
   library_t *libraries;
} librarylist_t;

symboltable_t vftr_read_symbols();

void vftr_symboltable_free(symboltable_t *symboltable_ptr);

#endif
