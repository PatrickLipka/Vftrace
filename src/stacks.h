#ifndef STACKS_H
#define STACKS_H

#include <stdbool.h>
#ifdef _DEBUG
#include <stdio.h>
#endif

#include "custom_types.h"
#include "symbol_types.h"
#include "stack_types.h"

int vftr_new_stack(int callerID, stacktree_t *stacktree_ptr,
                   symboltable_t symboltable, stack_kind_t stack_kind,
                   uintptr_t address, bool precise);

stacktree_t vftr_new_stacktree();

void vftr_stacktree_free(stacktree_t *stacktree_ptr);

void vftr_finalize_stacktree(stacktree_t *stacktree_ptr);

void vftr_print_stacktree(FILE *fp, stacktree_t stacktree);

void vftr_print_stacklist(FILE *fp, stacktree_t stacktree);

#endif
