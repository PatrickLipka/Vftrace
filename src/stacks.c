#include <stdlib.h>
#include <stdbool.h>
#ifdef _DEBUG
#include <stdio.h>
#endif

#include "stacks.h"
#include "symbols.h"
#include "search.h"

void vftr_stacktree_realloc(stacktree_t *stacktree_ptr) {
   stacktree_t stacktree = *stacktree_ptr;
   while (stacktree.nstacks > stacktree.maxstacks) {
      int maxstacks = stacktree.maxstacks*1.4+2;
      stacktree.stacks = (stack_t*)
         realloc(stacktree.stacks, maxstacks*sizeof(stack_t));
      stacktree.maxstacks = maxstacks;
   }
   *stacktree_ptr = stacktree;
}

void vftr_stack_callees_realloc(stack_t *stack_ptr) {
   stack_t stack = *stack_ptr;
   while (stack.ncallees > stack.maxcallees) {
      int maxcallees = stack.maxcallees*1.4+2;
      stack.callees = (int*)
         realloc(stack.callees, maxcallees*sizeof(int));
      stack.maxcallees = maxcallees;
   }
   *stack_ptr = stack;
}

void vftr_insert_callee(int calleeID, stack_t *caller) {
   // append the callee to the list
   int idx = caller->ncallees;
   caller->ncallees++;
   vftr_stack_callees_realloc(caller);
   caller->callees[idx] = calleeID;
}

int vftr_new_stack(int callerID, stacktree_t *stacktree_ptr,
                   symboltable_t symboltable, stack_kind_t stack_kind,
                   uintptr_t address, bool precise) {
   int stackID = stacktree_ptr->nstacks;
   stacktree_ptr->nstacks++;
   vftr_stacktree_realloc(stacktree_ptr);

   stack_t *callerstack = stacktree_ptr->stacks+callerID;

   stack_t *stack = stacktree_ptr->stacks+stackID;
   stack->stack_kind = stack_kind;
   stack->address = address;
   stack->precise = precise;
   stack->caller = callerstack->lid;
   stack->maxcallees = 0;
   stack->ncallees = 0;
   stack->callees = NULL;
   stack->lid = stackID;
   stack->gid = -1;

   // search for the function in the symbol table
   int symbID = vftr_binary_search_symboltable(symboltable.nsymbols,
                                               symboltable.symbols,
                                               address);
   if (symbID >= 0) {
      stack->name = symboltable.symbols[symbID].name;
   } else {
      stack->name = "(UnknownFunctionName)";
   }

   //printf("Caller: %s, ", caller->name);
   //printf("Callee: %s\n", stack->name);
   vftr_insert_callee(stack->lid, callerstack);

   return stack->lid;
}

stack_t vftr_first_stack() {
   stack_t stack;
   stack.stack_kind = init;
   stack.address = (uintptr_t) NULL;
   stack.precise = true;
   stack.caller = -1;
   stack.maxcallees = 0;
   stack.ncallees = 0;
   stack.callees = NULL;
   stack.lid = 0;
   stack.gid = -1;
   stack.name = "init";
   return stack;
}

// recursively free the stack tree
void vftr_stack_free(stack_t *stacks_ptr, int stackID) {
   stack_t stack = stacks_ptr[stackID];
   if (stack.ncallees > 0) {
      for (int icallee=0; icallee<stack.ncallees; icallee++) {
         vftr_stack_free(stacks_ptr, stack.callees[icallee]);
      }
      free(stack.callees);
      stack.callees = NULL;
   } 
   stacks_ptr[stackID] = stack;
}

stacktree_t vftr_new_stacktree() {
   stacktree_t stacktree;
   stacktree.maxstacks = 0;
   stacktree.stacks = NULL;
   stacktree.nstacks = 1;
   vftr_stacktree_realloc(&stacktree);
   stacktree.stacks[0] = vftr_first_stack();
   return stacktree;
}

void vftr_stacktree_free(stacktree_t *stacktree_ptr) {
   stacktree_t stacktree = *stacktree_ptr;
   if (stacktree.nstacks > 0) {
      vftr_stack_free(stacktree.stacks, 0);
      free(stacktree.stacks);
      stacktree.stacks = NULL;
      stacktree.nstacks = 0;
      stacktree.maxstacks = 0;
   }
   *stacktree_ptr = stacktree;
}

#ifdef _DEBUG
void vftr_print_stack(FILE *fp, int level, stacktree_t stacktree, int stackid) {
   // first print the indentation
   for (int ilevel=0; ilevel<level; ilevel++) {
      fprintf(fp, "  ");
   }
   stack_t stack = stacktree.stacks[stackid];
   fprintf(fp, "%s (%llx): id=%d\n", stack.name,
           (unsigned long long) stack.address, stack.lid);
   level++;
   for (int icallee=0; icallee<stack.ncallees; icallee++) {
      vftr_print_stack(fp, level, stacktree, stack.callees[icallee]);
   }
}

void vftr_print_stacktree(FILE *fp, stacktree_t stacktree) {
   fprintf(fp, "Stacktree\n");
   vftr_print_stack(fp, 0, stacktree, 0);
}
#endif
