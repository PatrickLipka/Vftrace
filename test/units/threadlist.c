#include <stdlib.h>
#include <stdio.h>

#include "self_profile.h"
#include "symbol_types.h"
#include "symbols.h"
#include "stack_types.h"
#include "stacks.h"
#include "threads.h"

#include "dummysymboltable.h"

#ifdef _MPI
#include <mpi.h>
#endif

int main(int argc, char **argv) {
   INIT_SELF_PROF_VFTRACE;
#if defined(_MPI)
   PMPI_Init(&argc, &argv);
#else
   (void) argc;
   (void) argv;
#endif

   // dummy symboltable
   uintptr_t addrs = 123456;
   symboltable_t symboltable = dummy_symbol_table(6, addrs);

   // build stacktree
   stacktree_t stacktree = vftr_new_stacktree();

   char *name;
   int func1_idx = 0;
   name = vftr_get_name_from_address(symboltable, addrs+0);
   int func2_idx = vftr_new_stack(func1_idx, &stacktree,
                                  name, name, addrs+0, false);
   name = vftr_get_name_from_address(symboltable, addrs+1);
   int func3_idx = vftr_new_stack(func1_idx, &stacktree,
                                  name, name, addrs+1, false);
   name = vftr_get_name_from_address(symboltable, addrs+2);
   int func4_idx = vftr_new_stack(func3_idx, &stacktree,
                                  name, name, addrs+2, false);
   name = vftr_get_name_from_address(symboltable, addrs+3);
   int func5_idx = vftr_new_stack(func2_idx, &stacktree,
                                  name, name, addrs+3, false);
   name = vftr_get_name_from_address(symboltable, addrs+4);
   int func6_idx = vftr_new_stack(func2_idx, &stacktree,
                                  name, name, addrs+4, false);
   name = vftr_get_name_from_address(symboltable, addrs+5);
   int func7_idx = vftr_new_stack(func6_idx, &stacktree,
                                  name, name, addrs+5, false);

   threadtree_t threadtree = vftr_new_threadtree();
   int thread0_idx = 0;
   int thread1_idx = vftr_new_thread(thread0_idx, &threadtree);
   int thread2_idx = vftr_new_thread(thread0_idx, &threadtree);
   int thread3_idx = vftr_new_thread(thread2_idx, &threadtree);
   int thread4_idx = vftr_new_thread(thread0_idx, &threadtree);
   int thread5_idx = vftr_new_thread(thread3_idx, &threadtree);
   int thread6_idx = vftr_new_thread(thread4_idx, &threadtree);
   int thread7_idx = vftr_new_thread(thread2_idx, &threadtree);
   int thread8_idx = vftr_new_thread(thread1_idx, &threadtree);
   int thread9_idx = vftr_new_thread(thread0_idx, &threadtree);

   vftr_print_threadlist(stdout, threadtree);

   free_dummy_symbol_table(&symboltable);
   vftr_threadtree_free(&threadtree);
   vftr_stacktree_free(&stacktree);
#ifdef _MPI
   PMPI_Finalize();
#endif

   FINALIZE_SELF_PROF_VFTRACE;
   return 0;
}
