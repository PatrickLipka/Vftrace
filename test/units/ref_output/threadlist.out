   0: parent=-, lvl=0, subthreads=1,2,4,9
   1: parent=0, lvl=1, subthreads=8
   2: parent=0, lvl=1, subthreads=3,7
   3: parent=2, lvl=2, subthreads=5
   4: parent=0, lvl=1, subthreads=6
   5: parent=3, lvl=3
   6: parent=4, lvl=2
   7: parent=2, lvl=2
   8: parent=1, lvl=2
   9: parent=0, lvl=1
